libsolidity/parsing/Parser.cpp

```diff
@@ -1627,6 +1643,8 @@ Parser::LookAheadInfo Parser::peekStatementType() const
 	if (mightBeTypeName)
 	{
 		Token::Value next = m_scanner->peekNextToken();
+		if (Token::isElementaryTypeName(token) && Token::isStateMutabilitySpecifier(next, false))
+			return LookAheadInfo::VariableDeclaration;
```

Comment:
Which test uses this line?

Reply:
Ah I think most of them use it.

Reply:
Actually things like `address payable[` should resolve to `LookAheadInfo::IndexAccessStructure`. Basically I'm asking due to which reason can `address payable...` be handled differently than `address...` in this function?

Reply:
Is the fact that casts like `adress payable(x);` or `address payable[](x)` are not possible the only reason?

Reply:
If yes, then I think we should extend this function since those might be possible at some point and it could be rather hard to find the reason here.

Reply:
The main difference indeed is that at the moment ``address payable`` casts are disallowed. I'll look into extending this function, though.

Reply:
We may need to be able to look ahead two tokens in the scanner for that, though.

Reply:
As we just discussed, I added a comment here and I also added another test case ``syntaxTests/parsing/address_payable_type_expression.sol`` to show that ``address payable;`` as type expression is currently not allowed.